### PR TITLE
Add params serializer

### PR DESCRIPTION
**Pull request by markkaylor**
### What does it do?

- Adds the paramsSerializer to axios fetch client to serialize params object to a string using qs.stringify

### Why is it needed?

- To avoid string concatenation when building query params

### How to test it?

I've update everything below but not sure I got everything. Please let me know if you know some places I missed.

Check get, put, and post requests for all updated files. They should all still make the same requests with query params that are currently on main.

### TODO

- [x] Update packages/core/upload/admin/src/hooks/useAssets.js
- [x] Update packages/core/upload/admin/src/hooks/useFolder.js
- [x] Update packages/core/upload/admin/src/hooks/useFolders.js
- [x] Update packages/core/admin/admin/src/content-manager/pages/ListView/components/TableRows/index.js
- [x] Update packages/core/admin/admin/src/content-manager/components/CollectionTypeFormWrapper/index.js
- [x] Update packages/core/admin/admin/src/content-manager/components/SingleTypeFormWrapper/index.js
- [x] Update packages/core/admin/admin/src/hooks/useAdminUsers/useAdminUsers.js
- [x] Update packages/core/admin/admin/src/pages/AuthPage/components/Register/index.js
- [x] Update packages/core/admin/ee/admin/pages/SettingsPage/pages/AuditLogs/ListView/hooks/useAuditLogsData.js
- [x] Update packages/core/admin/ee/admin/pages/SettingsPage/pages/ReviewWorkflows/hooks/useReviewWorkflows.js

### CODE REVIEW

File: packages/core/admin/admin/src/content-manager/pages/ListView/utils/buildValidQueryParams.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+import createPluginsFilter from './createPluginsFilter';
+
+/**
+ * Creates a valid query params object
+ * This includes:
+ * - a filters clause
+ * - plugin options
+ */
+const buildValidQueryParams = (queryParams = {}) => {
```
Comment (markkaylor) [inline packages/core/admin/admin/src/content-manager/pages/ListView/utils/buildValidQueryParams.js:11]: ℹ️ This is the same as buildQueryString but returning an object instead of a string. buildQueryString will eventually be removed

Comment (soupette) [reply packages/core/admin/admin/src/content-manager/pages/ListView/utils/buildValidQueryParams.js:11]: Whach out for infinite loops when using an object, otherwise it's a good improvement!

File: packages/core/helper-plugin/src/utils/getFetchClient/index.js [hunk 2]
```diff
@@ -1,23 +1,34 @@
...
+
 const getFetchClient = (defaultOptions = {}) => {
   instance.defaults.baseURL = window.strapi.backendURL;
   return {
-    get: (url, config) => instance.get(normalizeUrl(url), { ...defaultOptions, ...config }),
+    get: (url, config) =>
+      instance.get(normalizeUrl(url), {
+        ...defaultOptions,
+        ...config,
+        paramsSerializer,
```
Comment (joshuaellis) [inline packages/core/helper-plugin/src/utils/getFetchClient/index.js:24]: I think(?) This might be able to be applied to the `instance`?

File: packages/core/admin/admin/src/content-manager/pages/ListView/utils/buildValidQueryParams.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    plugins: _,
+    _q: query,
+    ...otherQueryParams
+  } = {
+    ...queryParams,
+    ...createPluginsFilter(queryParams.plugins),
+  };
+
+  if (query) {
+    set(otherQueryParams, `_q`, encodeURIComponent(query));
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/pages/ListView/utils/buildValidQueryParams.js:26]: How come you feel the need to use `set`? I think we can avoid lodash here :)

Comment (markkaylor) [reply packages/core/admin/admin/src/content-manager/pages/ListView/utils/buildValidQueryParams.js:26]: This was just a quick and dirty copy paste of buildQueryString to get things working. I'll take a look at removing set.

File: packages/core/helper-plugin/src/utils/getFetchClient/index.js (hunk 2 above)
Comment (markkaylor) [reply packages/core/helper-plugin/src/utils/getFetchClient/index.js:24]: Yup sure can! Makes way more sense, thanks 👍

Comment (joshuaellis) [reply packages/core/helper-plugin/src/utils/getFetchClient/index.js:24]: amazing.

File: packages/core/admin/admin/src/content-manager/components/CollectionTypeFormWrapper/index.js [hunk 4]
```diff
@@ -268,7 +270,7 @@ const CollectionTypeFormWrapper = ({ allLayoutData, children, slug, id, origin }
         // Enable navigation and remove loaders
         dispatch(setStatus('resolved'));
 
-        replace(`/content-manager/collectionType/${slug}/${data.id}${rawQuery}`);
+        replace(`/content-manager/collectionType/${slug}/${data.id}`);
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/components/CollectionTypeFormWrapper/index.js:273]: I think this still needs the query otherwise when you edit a non-default locale you get redirected to the default locale :)

File: packages/core/admin/admin/src/content-manager/pages/ListView/components/TableRows/index.js [hunk 5]
```diff
@@ -53,7 +59,9 @@ export const TableRows = ({
   const handleCloneClick = (id) => async () => {
     try {
       const { data } = await post(
-        `/content-manager/collection-types/${contentType.uid}/auto-clone/${id}?${pluginsQueryParams}`
+        `/content-manager/collection-types/${contentType.uid}/auto-clone/${id}`,
+        {},
+        { params: { plugins: query?.plugins } || {} }
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/pages/ListView/components/TableRows/index.js:64]: won't this object always be true? is the `||` in the wrong place?

File: packages/core/admin/admin/src/content-manager/pages/ListView/index.js [hunk 6]
```diff
@@ -110,6 +110,61 @@ function ListView({
   const fetchClient = useFetchClient();
   const { post, del } = fetchClient;
 
+  const entityListQuery = useQuery(
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/pages/ListView/index.js:113]: Same here, i think this out to be a different PR, there's probably a lot more optimisations / better practice we could be doing in a focussed situation.

File: packages/core/admin/admin/src/content-manager/pages/ListView/utils/tests/buildValidGetParams.test.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    const params = buildValidGetParams(queryParams);
+
+    const expectedParams = {
+      page: '1',
+      pageSize: '10',
+      sort: 'name:ASC',
+      locale: 'en',
+    };
+
+    expect(params).toStrictEqual(expectedParams);
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/pages/ListView/utils/tests/buildValidGetParams.test.js:23]: what do you think about doing `matchInlineSnapshot` here instead?

File: packages/core/admin/admin/src/content-manager/pages/ListView/utils/tests/buildValidGetParams.test.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  });
+
+  it('should encode a search query', () => {
+    const _q = `test&query`;
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/pages/ListView/utils/tests/buildValidGetParams.test.js:27]: Isn't this already the encoded version? 🤔

File: packages/core/admin/admin/src/pages/AuthPage/components/Register/index.js [hunk 9]
```diff
@@ -61,10 +61,16 @@ const Register = ({ authType, fieldsToDisable, noSignin, onSubmit, schema }) =>
   useEffect(() => {
     if (registrationToken) {
       const getData = async () => {
+        console.log('register');
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/pages/AuthPage/components/Register/index.js:64]: rogue log :)

File: packages/core/admin/admin/src/content-manager/components/SingleTypeFormWrapper/index.js [hunk 10]
```diff
@@ -50,6 +50,50 @@ const SingleTypeFormWrapper = ({ allLayoutData, children, slug }) => {
   const { componentsDataStructure, contentTypeDataStructure, data, isLoading, status } =
     useSelector(selectCrudReducer);
 
+  useQuery(
```
[the remark this answers is not in the document]
Comment (markkaylor) [reply packages/core/admin/admin/src/content-manager/components/SingleTypeFormWrapper/index.js:53]: ok I can revert, but I think it might mean memoizing the params object

[the remark this answers is not in the document]
Comment (markkaylor) [reply packages/core/admin/admin/src/content-manager/components/SingleTypeFormWrapper/index.js:53]: At least I had an infinite loop and I believe that was the culprit. To be fair I didn't really try to solve the problem with the useEffect I just thought I would go ahead and convert to useQuery, but I agree it is out of scope for this PR

File: packages/core/admin/admin/src/content-manager/pages/ListView/components/TableRows/index.js (hunk 5 above)
Comment (markkaylor) [reply packages/core/admin/admin/src/content-manager/pages/ListView/components/TableRows/index.js:64]: Yeah thanks it was in the wrong place, but I think the check is unnecessary

File: packages/core/admin/admin/src/content-manager/pages/ListView/utils/buildValidGetParams.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    plugins: _,
+    _q: searchQuery,
+    ...validQueryParams
+  } = {
+    ...query,
+    ...createPluginsFilter(query.plugins),
+  };
+
+  if (searchQuery) {
+    validQueryParams._q = encodeURIComponent(searchQuery);
```
[the remark this answers is not in the document]
Comment (markkaylor) [reply packages/core/admin/admin/src/content-manager/pages/ListView/utils/buildValidGetParams.js:23]: Yes I believe we do, the paramsSerializer has encode to `false` to match how we were stringifying before. I could setup some logic over there to encode _q on params if you prefer.

File: packages/core/admin/admin/src/content-manager/components/SingleTypeFormWrapper/index.js [hunk 12]
```diff
@@ -187,17 +178,17 @@ const SingleTypeFormWrapper = ({ allLayoutData, children, slug }) => {
...
       }
     },
-    [del, slug, displayErrors, toggleNotification, searchToSend, dispatch, rawQuery]
+    [del, slug, displayErrors, toggleNotification, query, dispatch, rawQuery]
   );
 
   const onPost = useCallback(
     async (body, trackerProperty) => {
-      const endPoint = getRequestUrl(`${slug}${rawQuery}`);
+      const endPoint = getRequestUrl(`${slug}`);
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/SingleTypeFormWrapper/index.js:186]:
```suggestion
      const endPoint = getRequestUrl(slug);
```

File: packages/core/admin/admin/src/content-manager/components/SingleTypeFormWrapper/index.js [hunk 13]
```diff
@@ -287,18 +284,18 @@ const SingleTypeFormWrapper = ({ allLayoutData, children, slug }) => {
 
       return Promise.reject(err);
     }
-  }, [post, cleanReceivedData, displayErrors, slug, searchToSend, dispatch, toggleNotification]);
+  }, [post, cleanReceivedData, displayErrors, slug, query, dispatch, toggleNotification]);
 
   const onPut = useCallback(
     async (body, trackerProperty) => {
-      const endPoint = getRequestUrl(`${slug}${rawQuery}`);
+      const endPoint = getRequestUrl(`${slug}`);
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/SingleTypeFormWrapper/index.js:291]:
```suggestion
      const endPoint = getRequestUrl(slug);
```

File: packages/core/admin/admin/src/content-manager/pages/ListView/utils/buildValidGetParams.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/pages/ListView/utils/buildValidGetParams.js:23]: I think it's fine if you think we need it – maybe worth adding a comment just to remember this as I imagine a few people will have a similar question :)

**Review by gu-stav — COMMENTED** [review]
Very nice unification! Same as Josh: I'd prefer to discuss some data-fetching related changes in separate PRs, because the way it is done became very messy.

File: packages/core/admin/admin/src/content-manager/pages/ListView/utils/tests/buildValidGetParams.test.js [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    const params = buildValidGetParams(queryParams);
+
+    expect(params).toMatchInlineSnapshot(`
+      {
+        "locale": "en",
+        "page": "1",
+        "pageSize": "10",
+        "sort": "name:ASC",
+      }
+    `);
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/pages/ListView/utils/tests/buildValidGetParams.test.js:23]: nice use of inline snapshot ✅

File: packages/core/admin/admin/src/content-manager/components/SingleTypeFormWrapper/index.js [hunk 15]
```diff
@@ -50,6 +50,50 @@ const SingleTypeFormWrapper = ({ allLayoutData, children, slug }) => {
   const { componentsDataStructure, contentTypeDataStructure, data, isLoading, status } =
     useSelector(selectCrudReducer);
 
+  useQuery(
+    [slug, params],
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/SingleTypeFormWrapper/index.js:54]:
```suggestion
    ['content-manager', slug, params],
```

**Review by gu-stav — DISMISSED** [review]
One last minor comment, but other than that it looks very good and is IMO ready to go 🚀 Nice work!

**Review by joshuaellis — APPROVED** [review]
💯 great addition

File: packages/core/admin/admin/src/content-manager/components/SingleTypeFormWrapper/index.js (hunk 15 above)
Comment (markkaylor) [reply packages/core/admin/admin/src/content-manager/components/SingleTypeFormWrapper/index.js]: Actually this made me realize I forgot to remove useQuery when I reverted back to the original useEffect that was in place 🤦‍♂️ . I plan on opening a PR to migrate to useQuery for collection-types and single-types after this one is merged. So I'll apply this feedback there.

**Review by gu-stav — APPROVED** [review]
Even smaller - even better :)

**Review by gu-stav — APPROVED** [review]

**Review by joshuaellis — APPROVED** [review]

**Review by joshuaellis — APPROVED** [review]

**Comment by markkaylor** [comment]
I am still looking into the failing api integration tests.  Locally it's for upload, but here in gh actions it's something to do with review-workflows. Either way struggling to understand how it relates to these changes 🤷‍♂️.
